src/Analyzers/CSharp/CodeFixes/RemoveUnnecessarySuppressions/CSharpRemoveUnnecessaryNullableWarningSuppressionsCodeFixProvider.cs

```diff
@@ -3,50 +3,178 @@
+
+        var newRoot = FixAll(
+            document.Project.Solution.Services, root, diagnostics.Select(static d => d.AdditionalLocations[0].SourceSpan));
+
+        return document.WithSyntaxRoot(newRoot);
+    }
+
+    private static SyntaxNode FixAll(
+        SolutionServices services,
+        SyntaxNode root,
+        IEnumerable<TextSpan> spans)
     {
-        foreach (var diagnostic in diagnostics.OrderByDescending(d => d.Location.SourceSpan.Start))
+        var editor = new SyntaxEditor(root, services);
+
+        foreach (var span in spans.OrderByDescending(d => d.Start))
         {
-            if (diagnostic.AdditionalLocations[0].FindNode(getInnermostNodeForTie: true, cancellationToken) is PostfixUnaryExpressionSyntax unaryExpression)
+            if (root.FindNode(span, getInnermostNodeForTie: true) is PostfixUnaryExpressionSyntax unaryExpression)
             {
                 editor.ReplaceNode(
                     unaryExpression,
                     (current, _) => ((PostfixUnaryExpressionSyntax)current).Operand.WithTriviaFrom(current));
             }
         }
 
-        return Task.CompletedTask;
+        return editor.GetChangedRoot();
+    }
+
+    public override FixAllProvider? GetFixAllProvider()
+        => new RemoveUnnecessaryNullableWarningSuppressionsFixAllProvider();
+
+    private sealed class RemoveUnnecessaryNullableWarningSuppressionsFixAllProvider : FixAllProvider
+    {
+#if !CODE_STYLE
+        internal override CodeActionCleanup Cleanup => CodeActionCleanup.SyntaxOnly;
+#endif
+
+        public override async Task<CodeAction?> GetFixAsync(FixAllContext fixAllContext)
+        {
+            var cancellationToken = fixAllContext.CancellationToken;
+
+            // Fix-all for removing unnecessary `!` operators works in a fairly specialized fashion.  The core problem
+            // is that it's normal to have situations where a `!` operator is unnecessary in one linked document in one
+            // project, but necessary in another.  Consider something as mundane as `string.IsNullOrEmpty(s)`.  In
+            // projects that reference a modern, annotated, BCL, the nullable attributes on this method will allow the
+            // compiler to determine that `s` is non-null after the call, allowing superfluous `!` operators to be
+            // removed.  However, in projects that reference an unannotated BCL, no such determination can be made, and
+            // the `!` on a following statement may be necessary.
+            //
+            // To deal with this, we consider all linked documents together.  If a `!` operator is unnecessary in *all*
+            // linked documents, then we can remove it.  Otherwise, we must keep it.
+
+            var documentToDiagnostics = await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(fixAllContext).ConfigureAwait(false);
+
+            // Note: we can only do this if we're doing a fix-all in the solution level.  That's the only way we can see
```

Comment:
Could this be done in the regular fixer as well? It does seem like we're pretty likely to hit this. I think it can't be done in the analyzer b/c it doesn't know about linked docs, but could the regular fixer find them and check if the diagnostic was reported in the other contexts?

Reply:
yeah thinking about it more, I think there isn't an optimal answer for what to do with the regular code fix since the diagnostic will always be there (and would just be changing if the code fix is present or not).  so I don't feel super strongly about this.